pkg/kubelet/images/image_manager.go

```diff
@@ -84,52 +86,145 @@ func (m *imageManager) logIt(ref *v1.ObjectReference, eventtype, event, prefix,
+func (m *imageManager) RemoveAsyncPullsForPod(pod *v1.Pod) {
+	for _, container := range pod.Spec.InitContainers {
+		asyncKey := getAsyncPullKey(pod, &container)
+		if _, exists := m.asyncPulls.Load(asyncKey); exists {
+			m.asyncPulls.Delete(asyncKey)
+		}
 	}
 
+	for _, container := range pod.Spec.Containers {
+		asyncKey := getAsyncPullKey(pod, &container)
+		if _, exists := m.asyncPulls.Load(asyncKey); exists {
+			m.asyncPulls.Delete(asyncKey)
+		}
+	}
+}
+
+// ensureImageExistsInternal has logic shared between EnsureImageExists and EnsureImageExistsAsync
+// (imageRef, error message, error, imageSpec, pullRequired).
+func (m *imageManager) ensureImageExistsInternal(ref *v1.ObjectReference, pod *v1.Pod, container *v1.Container, pullSecrets []v1.Secret, podSandboxConfig *runtimeapi.PodSandboxConfig) (string, string, error, kubecontainer.ImageSpec, bool) {
+	logPrefix := fmt.Sprintf("%s/%s", pod.Name, container.Image)
+
 	// If the image contains no tag or digest, a default tag should be applied.
 	image, err := applyDefaultImageTag(container.Image)
 	if err != nil {
 		msg := fmt.Sprintf("Failed to apply default image tag %q: %v", container.Image, err)
 		m.logIt(ref, v1.EventTypeWarning, events.FailedToInspectImage, logPrefix, msg, klog.Warning)
-		return "", msg, ErrInvalidImageName
+		return "", msg, ErrInvalidImageName, kubecontainer.ImageSpec{}, false
 	}
 
 	spec := kubecontainer.ImageSpec{Image: image}
 	imageRef, err := m.imageService.GetImageRef(spec)
 	if err != nil {
 		msg := fmt.Sprintf("Failed to inspect image %q: %v", container.Image, err)
 		m.logIt(ref, v1.EventTypeWarning, events.FailedToInspectImage, logPrefix, msg, klog.Warning)
-		return "", msg, ErrImageInspect
+		return "", msg, ErrImageInspect, spec, false
 	}
 
 	present := imageRef != ""
 	if !shouldPullImage(container, present) {
 		if present {
 			msg := fmt.Sprintf("Container image %q already present on machine", container.Image)
 			m.logIt(ref, v1.EventTypeNormal, events.PulledImage, logPrefix, msg, klog.Info)
-			return imageRef, "", nil
+			return imageRef, "", nil, spec, false
 		}
 		msg := fmt.Sprintf("Container image %q is not present with pull policy of Never", container.Image)
 		m.logIt(ref, v1.EventTypeWarning, events.ErrImageNeverPullPolicy, logPrefix, msg, klog.Warning)
-		return "", msg, ErrImageNeverPull
+		return "", msg, ErrImageNeverPull, spec, false
+	}
+
+	return "", "", nil, spec, true
+}
+
+// EnsureImageExistsAsync pulls the image for the specified pod and container, and returns
+// (imageRef, error message, error).
+func (m *imageManager) EnsureImageExistsAsync(pod *v1.Pod, container *v1.Container, pullSecrets []v1.Secret, podSandboxConfig *runtimeapi.PodSandboxConfig) (string, error) {
+	logPrefix := fmt.Sprintf("%s/%s", pod.Name, container.Image)
+	ref, err := kubecontainer.GenerateContainerRef(pod, container)
+	if err != nil {
+		klog.Errorf("Couldn't make a ref to pod %v, container %v: '%v'", pod.Name, container.Name, err)
```

Comment:
ref is used in the ensureImageExistsInternal call below.
Why isn't err returned here ?

Reply:
Seems like it probably should, I took this from the existing code in `EnsureImageExists`. Do we have any context on why that didn't return an error?

Reply:
I think it is good practice to error out early, instead of waiting for ensureImageExistsInternal to fail.